server-common/src/main/java/org/apache/kafka/server/immutable/pcollections/PCollectionsImmutableMap.java

```diff
@@ -0,0 +1,223 @@
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        PCollectionsImmutableMap<?, ?> that = (PCollectionsImmutableMap<?, ?>) o;
+        return underlying().equals(that.underlying());
+    }
+
+    @Override
+    public int hashCode() {
+        return underlying().hashCode();
+    }
+
+    @Override
+    public V getOrDefault(Object key, V defaultValue) {
+        return underlying().getOrDefault(key, defaultValue);
+    }
+
+    @Override
+    public void forEach(BiConsumer<? super K, ? super V> action) {
+        underlying().forEach(action);
+    }
+
+    @Override
+    public void replaceAll(BiFunction<? super K, ? super V, ? extends V> function) {
+        // will throw UnsupportedOperationException; delegate anyway for testability
+        underlying().replaceAll(function);
+    }
+
+    @Override
+    public V putIfAbsent(K key, V value) {
+        // will throw UnsupportedOperationException; delegate anyway for testability
+        return underlying().putIfAbsent(key, value);
+    }
+
+    @Override
+    public boolean remove(Object key, Object value) {
+        // will throw UnsupportedOperationException; delegate anyway for testability
+        return underlying().remove(key, value);
+    }
+
+    @Override
+    public boolean replace(K key, V oldValue, V newValue) {
+        // will throw UnsupportedOperationException; delegate anyway for testability
+        return underlying().replace(key, oldValue, newValue);
+    }
+
+    @Override
+    public V replace(K key, V value) {
+        // will throw UnsupportedOperationException; delegate anyway for testability
+        return underlying().replace(key, value);
+    }
+
+    @Override
+    public V computeIfAbsent(K key, Function<? super K, ? extends V> mappingFunction) {
+        // will throw UnsupportedOperationException; delegate anyway for testability
+        return underlying().computeIfAbsent(key, mappingFunction);
+    }
+
+    @Override
+    public V computeIfPresent(K key, BiFunction<? super K, ? super V, ? extends V> remappingFunction) {
+        // will throw UnsupportedOperationException; delegate anyway for testability
+        return underlying().computeIfPresent(key, remappingFunction);
+    }
+
+    @Override
+    public V compute(K key, BiFunction<? super K, ? super V, ? extends V> remappingFunction) {
+        // will throw UnsupportedOperationException; delegate anyway for testability
+        return underlying().compute(key, remappingFunction);
+    }
+
+    @Override
+    public V merge(K key, V value, BiFunction<? super V, ? super V, ? extends V> remappingFunction) {
+        // will throw UnsupportedOperationException; delegate anyway for testability
+        return underlying().merge(key, value, remappingFunction);
+    }
+
+    @Override
+    public String toString() {
+        return "PCollectionsImmutableMap{" +
+            "underlying=" + underlying() +
```

Comment:
I think it would be more usable if we simply had the key/values here versus exposing the nested structure.